src/middleware/compression.rs

```diff
@@ -0,0 +1,420 @@
+        };
+        Ok(encoding.unwrap_or(self.default_encoding))
+    }
+
+    /// Consumes the response and returns an encoded version of it.
+    fn encode(&self, mut res: Response, encoding: Encoding) -> Response {
+        match encoding {
+            Encoding::Gzip => {
+                let mut body = Body::empty();
+                std::mem::swap(&mut body, res.body_mut());
+                let stream = stream::GzipEncoder::new(body, self.settings.gzip);
+                *res.body_mut() = Body::from_stream(stream);
+                res.headers_mut()
+                    .append(CONTENT_ENCODING, encoding.to_header_value());
+                res
+            }
+            Encoding::Deflate => {
+                let mut body = Body::empty();
+                std::mem::swap(&mut body, res.body_mut());
+                let stream = stream::ZlibEncoder::new(body, self.settings.deflate);
+                *res.body_mut() = Body::from_stream(stream);
+                res.headers_mut()
+                    .append(CONTENT_ENCODING, encoding.to_header_value());
+                res
+            }
+            Encoding::Brotli => {
+                let mut body = Body::empty();
+                std::mem::swap(&mut body, res.body_mut());
+                let stream = stream::BrotliEncoder::new(body, self.settings.brotli);
+                *res.body_mut() = Body::from_stream(stream);
+                res.headers_mut()
+                    .append(CONTENT_ENCODING, encoding.to_header_value());
+                res
+            }
+            Encoding::Zstd => {
+                let mut body = Body::empty();
+                std::mem::swap(&mut body, res.body_mut());
+                let stream = stream::ZstdEncoder::new(body, self.settings.zstd);
+                *res.body_mut() = Body::from_stream(stream);
+                res.headers_mut()
+                    .append(CONTENT_ENCODING, encoding.to_header_value());
+                res
+            }
+            Encoding::Identity => res,
+        }
+    }
+}
+
+impl<Data: Send + Sync + 'static> Middleware<Data> for Compression {
+    fn handle<'a>(&'a self, cx: Context<Data>, next: Next<'a, Data>) -> BoxFuture<'a, Response> {
+        box_async! {
+            let encoding = match self.preferred_encoding(cx.headers()) {
+                Ok(encoding) => encoding,
+                Err(e) => return e.into_response(),
+            };
+            let res = next.run(cx).await;
+            self.encode(res, encoding)
+        }
+    }
+}
+
+/// Middleware for handling incoming request decompression.
+///
+/// This middleware currently supports HTTP decompression under the `gzip`, `deflate`, `br`, and `zstd` algorithms.
+#[derive(Debug, Default)]
+pub struct Decompression {}
+
+impl Decompression {
+    /// Creates a new Decompression middleware.
+    pub fn new() -> Self {
+        Self {}
+    }
+
+    fn parse_encoding(s: &str) -> Result<Encoding, Error> {
```

Comment:
Should we use `TryInto<Encoding> for &str` instead of this?

Reply:
That would necessitate implementing `TryFrom<&str>` for `accept_encoding::Encoding`, and I'm not sure if I'm ready to go through all that work when the header value should be a valid UTF-8 string anyway :sweat_smile: